sr5824-msr9732-sm9453/project.md

```diff
@@ -0,0 +1,136 @@
+---
+title: "Interoperability in the Heterogeneous Cloud Environment"
+author:
+  - Sriram Ramesh <sriram.ramesh@nyu.edu>
+  - Metarya Ruparel <msr9732@nyu.edu>
+  - Safwan Mahmood <sm9453@nyu.edu>
+---
+
+For our Big Data & Machine Learning final project we plan to hypothesize a solution
+for **Interoperability in the Heterogeneous Cloud Environment**, a blatant issue 
+industry wide.
+
+# Motivation
+
+After the burst of cloud platforms in 2006 with the introduction of Amazon
+Web Services, a lot of companies started migrating their services to the 
+cloud platform from a bare metal lying in one of their own datacenters.
+Big hulking datacenters are a necessary evil, a barely concealed secret. 
+No one wants to build one. They’re certainly impressive – they resemble the 
+set of a science fiction movie, row upon row of silent machines. But they’re 
+wildly expensive, and their maintenance challenges are never ending. 
+
+Cloud Platforms provided these companies the ability to access shared, online computing 
+resources, without the over head of maintaining them. However, these providers often 
+offer their own proprietary applications, interfaces, APIs and infrastructures, which
+somewhat makes these providers unique. Due to the uniqueness (be it performance or costing)
+companies would want to onboard different set of services onto various providers.
+For example, Salesforce.com, which  started the wave of SaaS back in the late 1900's 
+, has only recently started it's migration from 1st party datacenters to these 3rd party
+providers. Some of their micro-services are being on-boarded to GCP, while a couple of
+others to AWS, resulting in a heterogeneous cloud environment. This heterogeneous 
+environment makes it difficult for companies to change cloud service providers; in 
+any case. Exploring capabilities to support the automated migration from one provider 
+to another is an active, open research area. A lot of people have been pursuing approaches 
+to reduce the impact of vendor lock-in by investigating the cloud migration problem 
+at the level of the VM. However, the migration downtime, decoupling VM from
+underlying systems and security of live channels remain open issues and as a part of
+this project we aim to hypothesize a solution to at least one blatant issue in the
+process of migration.
+
+# Proposed Solution
+
+As we are moving towards heterogeneous cloud, we want to write Cloud-Agnostic
+applications. To make the transition easier, we propose a static code analyser
+to find code snippets specific to a cloud provider using SonarQube Quality
```

Comment:
When you say "cloud provider specific code snippets" what do you mean: do you mean calls through libraries like Boto that set up infrastructure, do you mea APIs to access things like S3, do you mean something else?

What are you assuming about this code? Static analysis is not magic, so are you assuming you know the libraries a-priori, that different cloud providers have different code paths, something else?

Reply:
When we say cloud specific code snippets, we mean the core API calls for a specific language, in our case Java. Also, your example of S3 is one of them.

In the initial setting of the tool, we assume we know the code snippets and paths of the cloud providers a-priori (This is readily available in the documentaions of the cloud providers. Ex: For S3. Buckets on AWS https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/examples-s3-buckets.html ). We will use this knowledge base in aiding our static code analysis with SonarQube.